website/src/documentation/runtime/environments.md

```diff
@@ -0,0 +1,187 @@
+
+```
+$USER-docker-apache.bintray.io/beam
+```
+
+When you [build SDK container images](#building-container-images), you can override the default version and location.
+
+To specify an older Python SDK version, like 2.3.0, build the container with the `docker-tag` option:
+
+```
+./gradlew docker -Pdocker-tag=2.3.0
+```
+
+To change the `docker` target, build the container with the `docker-repository-root` option:
+
+```
+./gradlew docker -Pdocker-repository-root=$LOCATION
+```
+
+## Customizing container images
+
+You can add extra dependencies or serialization files to container images so the execution engine doesn't need them.
+
+To customize a container image, either:
+* [Write a new](#writing-new-dockerfiles) [Dockerfile](https://docs.docker.com/engine/reference/builder/) on top of the original
+* [Modify](#modifying-dockerfiles) the [original Dockerfile](https://github.com/apache/beam/blob/master/sdks/python/container/Dockerfile) and reimage the container
+
+It's often easier to write a new Dockerfile, but you can customize anything, including the base OS, by modifying the original.
+
+### Writing new Dockerfiles on top of the original {#writing-new-dockerfiles}
+
+<ol>
+    <li>
+        Pull a <a href="https://console.cloud.google.com/gcr/images/apache-beam-testing/GLOBAL/beam/sdks/release">prebuilt SDK container image</a> for your target language and version.
+    </li>
+    <li>
+        <a href="https://docs.docker.com/develop/develop-images/dockerfile_best-practices/">Write a new Dockerfile</a> that <a href="https://docs.docker.com/engine/reference/builder/#from">designates</a> the original as its <a href="https://docs.docker.com/glossary/?term=parent%20image">parent</a>
+    </li>
+    <li>
+        Build a child image: <pre>docker build -f /path/to/new/Dockerfile</pre>
+    </li>
+</ol>
+
+### Modifying the original Dockerfile {#modifying-dockerfiles}
+
+1. Pull the [prebuilt SDK container image](https://console.cloud.google.com/gcr/images/apache-beam-testing/GLOBAL/beam/sdks/release) for your target language and version
+2. Customize the [Dockerfile](https://github.com/apache/beam/blob/master/sdks/python/container/Dockerfile). If you're adding dependencies from [PyPI](https://pypi.org/), use `base_image_requirements.txt` instead of the Dockerfile.
+3. [Reimage](#building-container-images) the container
+
+### Testing customized images
+
+Test customized images with PortableRunner. PortableRunner targets the default Docker images, like `${USER}-docker-apache.bintray.io/beam/python${version}:latest`. You can specify a different container with the `--environment_config` flag.
+
+#### Testing on local machines
+
+```
+python -m apache_beam.examples.wordcount --input /path/to/inputfile --output /path/to/write/counts \
+--runner=PortableRunner \
+--job_endpoint=embed \
+[--environment_config={docker_image_name}]
+```
+
+#### Testing on local Flink clusters
```

Comment:
Testing on local and dataflow vs testing on Flink and Spark cluster don't share the same format.
Let's make it consistent, all use code blocks or all use pre blocks.